### ISSUE TITLE
mono-tools packages broken in preview-xenial repo

### ISSUE THREAD

**Issue by SwooshyCueb** [mono/mono#7285]
I was directed to post this issue here, rather than the `mono-tools` or `linux-packaging-mono-tools` repo.

The `gendarme`, `mono-tools-devel` and `mono-tools-gui` packages in the `preview-xenial` package repository depend on `libmono-cecil-private-cil` 5.2, and therefore will not install. It looks like they depend on whatever version of `libmono-cecil-private-cil` is present on the system when the packages are built. I think this issue exists in the `stable-xenial` repository as well. It looks like it has been like this for some time.

**Comment by KeirDavis1** [comment]
Mono 5.8.0 works fine, however upgrading to 5.10.0.140 breaks Sonarr & Jackett. Afraid to upgrade my Radarr and Lidarr VMs in case this will break those too.

**Comment by rrd911** [comment]
Someone on the Sonarr Git posted this. I had to remove a bunch of dependencies but was finally able to get things working again.
Pin your version of mono to 5.8.0.127 for the time being.

I did something like this (on Ubuntu):

apt remove mono-complete mono-devel (or whatever you installed for mono, I did mono-complete)
apt autoremove

I edited /etc/apt/sources.list.d/mono-xamarin.list to read:

deb http://download.mono-project.com/repo/ubuntu xenial/snapshots/5.8.0.127 main

Note in the above line, I locked to the 5.8.0.127 version, which still works with sonarr/radarr/jackett.

Then I did an apt update.

Then I did an apt install mono-complete, that got me back down to 5.8.0.127.

Hope this helps someone.

**Comment by albertcard** [comment]
Can confirm. After a moment of panic due to the upgrade causing sonarr/radarr/jackett to no longer start, @rrd911's solution solved the issue.

**Comment by marek-safar** [comment]
@directhex this looks like packaging regression, could you investigate

**Comment by directhex** [comment]
mono-tools contains three pieces which build against Cecil - `minvoke`, `gui-compare`, and `Gendarme`.

The Debian packaging is patched to use a distribution-provided standalone copy of Cecil (not the version bundled with Mono), and to bundle that standalone copy with the mono-tools executables - this is because historically Mono's bundled Cecil has had poor ABI stability, or is patched where the version number reported in no way reflects the ABI.

There is a bug in the package tooling where it detects the assemblyref on `Mono.Cecil.dll`, but resolves the Mono copy, not the private copy, because the assemblyref signatures match.

There is a further bug in the package tooling where the function to manually ignore bad dependencies (via an overrides file) does not work.

The fix requires:
1. A patch to `dh_clideps` in `cli-common-dev` to fix loading of `clideps-override` files
2. A patch to `mono-tools` to include clideps-overrides files, which skip the spurious dependency on Cecil.

**Comment by Taloth** [comment]
@KeirDavis1 @rrd911 @albertcard Jumping the gun there a bit. The failure of some apps is definitely not a bug in Mono, but rather NLog mishandling a new implicit cast operator that got introduced in the .net framework. It's been presumably fixed in NLog 4.5 release candidates a while ago, but not final yet. Please track your respective app github issue to get notified once a fix is available.

Lets leave the further discussion in this github issue to the unrelated mono-tools packaging problem.

**Comment by directhex** [comment]
Bad:
```
Source: mono-tools
Version: 4.2-2xamarin5+ubuntu1604b1
Depends: mono-runtime (>= 3.0~), libmono-cecil-private-cil (>= 5.2.0.196), libmono-cecil-private-cil (<< 5.2.0.197), libmono-corlib4.5-cil (>= 4.0.0~alpha1), libmono-sharpzip4.84-cil (>= 1.0), libmono-system-core4.0-cil (>= 4.0.0~alpha1), libmono-system-drawing4.0-cil (>= 3.0.6), libmono-system-web-services4.0-cil (>= 1.0), libmono-system-windows-forms4.0-cil (>= 1.0), libmono-system-xml4.0-cil (>= 3.12.0), libmono-system4.0-cil (>= 4.0.0~alpha1)
```

Good:
```
Source: mono-tools
Version: 4.2-2xamarin7+ubuntu1604b1
Depends: mono-runtime (>= 3.0~), libmono-corlib4.5-cil (>= 4.0.0~alpha1), libmono-sharpzip4.84-cil (>= 1.0), libmono-system-core4.0-cil (>= 4.0.0~alpha1), libmono-system-drawing4.0-cil (>= 3.0.6), libmono-system-web-services4.0-cil (>= 1.0), libmono-system-windows-forms4.0-cil (>= 1.0), libmono-system-xml4.0-cil (>= 3.12.0), libmono-system4.0-cil (>= 4.0.0~alpha1)
```

Published in `preview-*` repos, currently republishing to `stable-*` and `vs-*`
